README.md
# HalfHeight

An extension to `UI.NCurses` for drawing half-height console graphics.

By drawing a Unicode [Upper Half Block](https://www.compart.com/en/unicode/U+2580) character (▀) with different foreground and background colors, one can simulate drawing two rows of a 2D grid of colors in one row, achieving grid cells that take up half a console row alongside regular console text.

![](example.png)

There's a [brief writeup](https://askham.ai/2020/11/26/half-height-console-graphics.html) on my blog.

The main caveat is that one is now restricted to 15 unique colors: `UI.NCurses` usually lets you define 255 custom foreground/background combinations, but now we might have any two colors appear as our foreground/background. 16 colors would tip us over, requiring 256 possible pairs, so we are limited to 15 colors and 225 pairings.

## Usage Example

``` haskell
greyScale :: [String]
greyScale =
  [ "000000",
    "111111",
    -- ... 12 intermediate hex strings omitted ...
    "ffffff"
  ]


-- Calculate the Mandelbrot steps-to-divergence and plot as Color 1 to Color 15
mandelbrot :: (Fractional a, RealFloat a) => a -> a -> Color
mandelbrot x y = Color (16 - ((+) 1 $ steps (0 :+ 0) (x' :+ y') 0))
  where
    x' = (x - 100) * 0.025
    y' = (y - 50) * 0.025
    steps z c i
      | i > 13 || magnitude z' > 2 = i
      | otherwise = steps z' c (i + 1)
      where
        z' = z * z + c

-- Draw the Mandelbrot set to the buffer point by point
drawMandelbrot :: Buffer -> Buffer
drawMandelbrot buffer =
  foldl'
    (\b (x, y) -> setXY x y (mandelbrot (fromIntegral x) (fromIntegral y)) b)
    buffer
    [(x, y) | x <- [0 .. 149], y <- [0 .. 99]]

main :: IO ()
main = do
  -- Register the 15 greyscale hex colors as Color 1 through Color 15.
  -- Every combination of these colors (fg and bg) is registered with Curses
  -- with a unique ID, which we return in the colorMap.
  colorMap <- runCurses $ initHexColors greyScale

  -- Creates a new buffer and draws the Mandelbrot set in each cell using the
  -- greyscale colors 1 through 15
  let buffer = drawMandelbrot $ mkBuffer 150 100 (Color 1)

  runCurses $ do
    -- Hide the cursor
    setEcho False
    setCursorMode CursorInvisible

    -- Enter an infinite drawing loop displaying the buffer.
    forever $ do
      let drawOp = do
            -- First draw the buffer to the screen.
            drawBuffer colorMap 0 0 buffer
            -- Now draw some text in the centre.
            -- We set the color to curses ID corresponding to fg 15 (white) and bg 1 (black)
            let (Just c) = colorId colorMap (Color 15) (Color 1)
            setColor c
            moveCursor 25 50
            drawText "| Text demonstrating that each cell is half the row height |"
      w <- defaultWindow
      updateWindow w drawOp
      render
```

In this example we:

- Register `Color 1` through `Color 15` by providing a list of 15 hex strings to `initHexColors`. This sets the terminal colors appropriately, registers a unique curses `ColorID` for every possible foreground/background combination of each of the 15 colors, and returns a map from (fg, bg) tuples to corresponding `ColorID`.
- Create a `Buffer` (a 2D `Vector` of `Color`) by providing dimensions and a default color to `mkBuffer`.
- Draw on this buffer using `SetXY`.
- When ready to blit to the screen, `drawBuffer` will create the appropriate curses `Update ()` operations for writing the corresponding Upper Half Block characters to the terminal to render the buffer rows at half-height.


ChangeLog.md
# Changelog for halfheight

## Unreleased changes


src/UI/NCurses/HalfHeight.hs
module UI.NCurses.HalfHeight
  ( ColorMap (),
    colorId,
    initHexColors,
    Buffer (),
    mkBuffer,
    setXY,
    drawBuffer,
  )
where

import Control.Applicative (ZipList (ZipList, getZipList))
import Data.Int (Int16)
import qualified Data.IntMap.Strict as IM
import Data.List.Split as LS (chunksOf)
import qualified Data.Map.Strict as M
import Data.Maybe (catMaybes)
import Numeric (readHex)
import UI.NCurses
  ( Color (Color),
    ColorID,
    Curses,
    Glyph (Glyph),
    Update,
    defineColor,
    drawGlyph,
    moveCursor,
    newColorID,
    setColor,
  )

toTuple3 :: [a] -> Maybe (a, a, a)
toTuple3 [a, b, c] = Just (a, b, c)
toTuple3 _ = Nothing

-- A pair of colors s.t. we can display half-characters using unicode blocks.
-- Represents the foreground and background color of a combined half-height cell.
-- Color must be one of Curses Color 1 to Color 15.
type ColorPair = (Int16, Int16)

-- | A map to store the registered color pairings.
-- Keyed by fg*100 + bg for speed.
type ColorMap = IM.IntMap ColorID

colorKey :: ColorPair -> Int
colorKey (fg, bg) = fromIntegral $ fg * 100 + bg

-- | Get the registered curses ID for a given foreground / background colour combination.
colorId :: ColorMap -> Color -> Color -> Maybe ColorID
colorId colors (Color fg) (Color bg) = IM.lookup (colorKey (fg, bg)) colors

-- Takes a hex code and converts to a tuple of RGB 1000 values.
-- These are required by the curses color register.
parseColor :: String -> Maybe (Integer, Integer, Integer)
parseColor hex = toTuple3 scaled
  where
    rgbVals = fst <$> (head . readHex <$> LS.chunksOf 2 hex)
    scaled = (\x -> (x * 1000) `div` 255) <$> rgbVals

-- Defines as many custom colors as we have in hexColors.
-- Supports thousands but we can ultimately only use 255.
-- By doubling vertical resolution we can only use <16 since any color can
-- appear next to any color.
registerHexColors :: [String] -> Curses ()
registerHexColors hexes = sequence_ . getZipList $ ZipList mkColors <*> ZipList colorIDs
  where
    colorIDs = [1 ..] :: [Int16]
    mkColors =
      [ \i -> defineColor (Color i) r g b
        | (r, g, b) <- catMaybes $ parseColor <$> hexes
      ]

-- | Takes all custom colors and defines + persists them as an overlap map.
-- Registers the given list of 15 hex colors with Curses.
-- Stores each combination of colors (foreground and background) as a unique ColorID
-- Returns a map from (fg, bg) pairs to the corresponding registered ColorID for the pair.
initHexColors :: [String] -> Curses ColorMap
initHexColors hexes = do
  registerHexColors hexes
  sequenceA $ IM.fromList (zip colorKeys newColors)
  where
    colorIxs = [1 .. 15]
    colors = Color <$> colorIxs
    colorIDs = [1 .. length colors ^ 2]
    colorCombos = (,) <$> colors <*> colors
    colorKeys = colorKey <$> ((,) <$> colorIxs <*> colorIxs)
    colorCreators = ZipList $ uncurry newColorID <$> colorCombos
    newColors = getZipList $ colorCreators <*> ZipList (fromIntegral <$> colorIDs)

-- | Representation of a 2D grid of half-height coloured cells.
-- Stored in a map with the original coordinates for convenience + speed.
-- Keys are 10000x + y since buffers rarely get this large.
type Buffer = IM.IntMap ((Int, Int), ColorPair)

bufferKey :: (Int, Int) -> Int
bufferKey (x, y) = 10000 * x + y

-- | Make an empty buffer with the given color set as the background.
mkBuffer :: Int -> Int -> Color -> Buffer
mkBuffer width height (Color bgCol) =
  IM.fromList
    [(bufferKey (x, y), ((x, y), (bgCol, bgCol))) | y <- [0 .. (height `div` 2) -1], x <- [0 .. width -1]]

-- | Set a single block in the graphics buffer to the given color.
setXY :: Int -> Int -> Color -> Buffer -> Buffer
setXY x y (Color c) =
  IM.adjust
    ( \(p, (c1, c2)) -> case y `rem` 2 of
        0 -> (p, (c, c2))
        1 -> (p, (c1, c))
    )
    (bufferKey (x, y `div` 2))

-- | Draws the graphics buffer to the screen at the given top-left position.
-- Must first have obtained a color map by registering colors with initHexColors.
drawBuffer :: ColorMap -> Int -> Int -> Buffer -> Update ()
drawBuffer colorMap topLeftX topLeftY buffer = do
  let drawOne (_, ((x, y), (fg, bg))) =
        case colorId colorMap (Color fg) (Color bg) of
          Just cId -> do
            moveCursor (fromIntegral $ y + topLeftY) (fromIntegral $ x + topLeftX)
            setColor cId
            drawGlyph $ Glyph '▀' []
          Nothing -> return ()
  sequence_ $ drawOne <$> IM.toList buffer


app/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Monad (forever)
import Control.Monad.IO.Class (MonadIO (liftIO))
import Control.Monad.State.Strict
  ( MonadState (get, put),
    execStateT,
    forever,
    lift,
    runStateT,
  )
import Data.Complex (Complex (..), magnitude)
import Data.List (foldl')
import qualified Data.Text as T
import System.Random (Random (randomR), newStdGen)
import UI.NCurses
  ( Color (..),
    CursorMode (..),
    defaultWindow,
    drawText,
    moveCursor,
    render,
    runCurses,
    setColor,
    setCursorMode,
    setEcho,
    updateWindow,
  )
import UI.NCurses.HalfHeight
  ( Buffer,
    colorId,
    drawBuffer,
    initHexColors,
    mkBuffer,
    setXY,
  )

-- Set up a 15-color greyscale pallette
greyScale :: [String]
greyScale =
  [ "000000",
    "111111",
    "222222",
    "333333",
    "444444",
    "666666",
    "777777",
    "888888",
    "999999",
    "aaaaaa",
    "bbbbbb",
    "cccccc",
    "dddddd",
    "eeeeee",
    "ffffff"
  ]

-- Calculate the Mandelbrot steps-to-divergence and plot as Color 1 to Color 15
mandelbrot :: (Fractional a, RealFloat a) => a -> a -> Color
mandelbrot x y = Color (16 - ((+) 1 $ steps (0 :+ 0) (x' :+ y') 0))
  where
    x' = (x - 100) * 0.025
    y' = (y - 50) * 0.025
    steps z c i
      | i > 13 || magnitude z' > 2 = i
      | otherwise = steps z' c (i + 1)
      where
        z' = z * z + c

-- Draw the Mandelbrot set to the buffer point by point
drawMandelbrot :: Buffer -> Buffer
drawMandelbrot buffer =
  foldl'
    (\b (x, y) -> setXY x y (mandelbrot (fromIntegral x) (fromIntegral y)) b)
    buffer
    [(x, y) | x <- [0 .. 149], y <- [0 .. 99]]

mandelbrotMain :: IO ()
mandelbrotMain = do
  -- Register the 15 greyscale hex colors as Color 1 through Color 15.
  -- Every combination of these colors (fg and bg) is registered with Curses
  -- with a unique ID, which we return in the colorMap.
  colorMap <- runCurses $ initHexColors nord

  -- Creates a new buffer and draws the Mandelbrot set in each cell using the
  -- greyscale colors 1 through 15
  let buffer = drawMandelbrot $ mkBuffer 150 100 (Color 1)

  runCurses $ do
    -- Hide the cursor
    setEcho False
    setCursorMode CursorInvisible

    -- Enter an infinite drawing loop displaying the buffer.
    forever $ do
      let drawOp = do
            -- First draw the buffer to the screen.
            drawBuffer colorMap 0 0 buffer
            -- Now draw some text in the centre.
            -- We set the color to curses ID corresponding to fg 15 (white) and bg 1 (black)
            let (Just c) = colorId colorMap (Color 15) (Color 1)
            setColor c
            moveCursor 25 50
            drawText "| Text demonstrating that each cell is half the row height |"
      w <- defaultWindow
      updateWindow w drawOp
      render

nord :: [String]
nord =
  [ "3b4252",
    "bf616a",
    "a3be8c",
    --"ebcb8b",
    "81a1c1",
    "b48ead",
    "88c0d0",
    "e5e9f0",
    "4c566a",
    "bf616a",
    "a3be8c",
    "ebcb8b",
    "81a1c1",
    "b48ead",
    "8fbcbb",
    "eceff4"
  ]

performanceMain :: IO ()
performanceMain = do
  colorMap <- runCurses $ initHexColors greyScale
  -- Precompute the buffers
  let buffer = mkBuffer 100 100 (Color 1)
      col x y t =
        let x' = fromIntegral (x - 50)
            y' = fromIntegral (y - 50)
            t' = fromIntegral t
         in abs $ round $ (14 * ((sin $ 0.1 * x') * (cos $ 0.1 * y')) / (sin $ 0.1 * t')) + 1
      nextBuf b t = foldl' (\b (x, y) -> setXY x y (Color $ col x y t) b) b [(x, y) | x <- [0 .. 99], y <- [0 .. 99]]
      bufs = cycle $ scanl nextBuf buffer [0 .. 63]
  runCurses $ do
    _ <- (flip runStateT) (bufs, 0) $ do
      lift $ setEcho False
      lift $ setCursorMode CursorInvisible
      forever $ do
        (bufs, t) <- get
        let b = head bufs
            drawOp = do
              drawBuffer colorMap 0 0 b
              let (Just c) = colorId colorMap (Color 15) (Color 1)
              setColor c
              moveCursor 0 0
              drawText $ T.pack $ show t
        w <- lift defaultWindow
        lift $ updateWindow w drawOp
        lift render
        put (tail bufs, t + 1)
    return ()

main :: IO ()
main = do
  --mandelbrotMain
  performanceMain
